### DISCUSSION TITLE
Using xDrip plus with CamAPS

### DISCUSSION THREAD

**Discussion by ShaggydogUK** [NightscoutFoundation/xDrip#2008] [Q&A]
Hi all, I've tried to setup xDrip with the alternate slot number so I can use it with CamAPS.  It's not been successful, as even though I've put xDrip on slot 2 instead of the receiver that I normally have there as soon as xDrip is giving values then CamAPS (on the same phone) stops.  I'm using Android 9 on an LG G6.

I've since deleted xDrip, and although CamAPS has started giving values again my receiver hasn't.

I uploaded the logs from xDrip earlier today.

Does anyone have any ideas, or is sharing the 2 slots with one phone just not going to work?

Cheers

Simon

**Comment by Navid200** [comment]
Can you show a screenshot of the system status page?
You can blackout the serial number if you like.  But, please leave everything else visible.

**Reply by ShaggydogUK** [reply]
<img width="350" src="https://user-images.githubusercontent.com/30231246/155189997-581e35dd-713f-45aa-8633-c2201270b013.png" />

I've reinstalled - as soon as xDrip connected to the transmitter I lost contact with CamAPS

**Reply by Navid200** [reply]
Please swipe left to go to the G5/G6 status page and post that one too.

**Comment by ShaggydogUK** [comment]
<img width="350" src="https://user-images.githubusercontent.com/30231246/155197857-1a8311c5-16fa-4332-9ebd-8f22b53a5016.png" />
Sorry - forgot about that!

**Reply by Navid200** [reply]
There is a start command stuck in the queue.
You can clear the queue:  https://navid200.github.io/xDrip/docs/Clear-queue.html

**Reply by Navid200** [reply]
Just to clarify:
There are only 2 slots.  If you are going to keep the receiver on, you will be occupying slot 1 by the receiver.
That will leave slot 2 for CamAPS.  At that point, it makes no difference what slot number you assign to xDrip because you have already occupied both slots.

**Reply by sboernemeyer** [reply]
what do you think:
possible to fix ?
@Nacid200: orherwise - is it possible building a Dexcom Clarity bridge for CamAPS FX ?
THX

**Reply by Navid200** [reply]
There is nothing to fix.  I warned everyone when I suggested this test that it may not work since both apps are on the same phone.

I don't know exactly how the pump is controlled by CamAPS.  May be we can add similar control to xDrip.

I'm not the right person to advise.  I don't use a pump.  So, I don't know what the alternatives are.  For example, I don't know how difficult it is to use AAPS or any other alternative.
If it's not too difficult, there is no need to change anything other than stop using CamAPS and switch to AAPS.  
Of course, you may have a very justifiable reason for not wanting to do that.
This is what makes things complicated here.  There are so many different requests, and we have to decide which one is absolutely necessary.
There is no way we can address all requests.

**Reply by ShaggydogUK** [reply]
@Navid200 you guys do an amazing job, and as Dexcom/CamAPS are working on getting CamAPS values through to the Dexcom servers it hopefully won't be necessary in the near(-ish - the link has been promised for over a year but Dexcom seem to be dragging their feet) future.  When that happens we will be able to link to the Nightscout family with ease.

**Reply by webwude** [reply]
> @Navid200 you guys do an amazing job, and as Dexcom/CamAPS are working on getting CamAPS values through to the Dexcom servers it hopefully won't be necessary in the near(-ish - the link has been promised for over a year but Dexcom seem to be dragging their feet) future.

Where did you get this from? Are there really plans to link Camaps with Dexcom?

**Reply by ShaggydogUK** [reply]
> > @Navid200 you guys do an amazing job, and as Dexcom/CamAPS are working on getting CamAPS values through to the Dexcom servers it hopefully won't be necessary in the near(-ish - the link has been promised for over a year but Dexcom seem to be dragging their feet) future.
> 
> Where did you get this from? Are there really plans to link Camaps with Dexcom?

CamAPS already use Dexcom, but CamAPS are hoping to get Dexcom values available on Clarity and Dexcom Share which will allow readings to be drawn through to Nightscout.  I'm not using CamAPS anymore so I don't know if that has happened yet - as I say, the CamAPS team are ready for it, but it is down to Dexcom to make the changes necessary at their end - which, last I knew, they hadn't done.   The information that it was going to happen came from CamAPS, they are as impatient as well!

**Comment by ShaggydogUK** [comment]
Yes, I realise that - the receiver is turned off when I'm attempting to connect xDrip - which I'm uninstalling when CamAPS stops connecting, so I'm reinstalling each time.  Here's the screenshot from the latest install

<img width="350" src="https://user-images.githubusercontent.com/30231246/155216913-2825d708-72b8-46aa-9762-4c95a06ee911.png" />

**Reply by Navid200** [reply]
And you have set the slot number to 1?

**Reply by ShaggydogUK** [reply]
Yes.  It seems that either one will connect, but not both.  Usually it is xDrip that connects, except the previous time when CamAPS stayed connected while the start command failed to run.

**Reply by Navid200** [reply]
OK,
Thanks for trying.
Of course, you can use a second phone.  That will solve the problem.  But, of course, you don't want to have to carry two phones.
I wonder if there is a way to create the illusion that xDrip is installed on another phone when in reality it is not.

Anyway, thanks for testing and sharing the results.

**Reply by ShaggydogUK** [reply]
Thanks for trying to make it work - it's good to see that we CamAPSers are forgotten about!

Simon

**Reply by ShaggydogUK** [reply]
I have a smart watch - can i use that as a collector from slot 1?  Also, is it possible that using a bluetooth dongle on the phone would enable it to act as 2 devices?  Just thoughts that have come to mind!

I've just plugged a Bluetooth dongle into my phone and it wasn't recognised :-(  Firmware of watch not supported for patching, so that option goes out of the windows as well.  Oh well, back to the receiver!

**Comment by tolot27** [comment]
As long as you use the Dexcom transmitter, only two devices/apps can be connected at a time with the transmitter. In your case, it is your pump (which one do you have) and CamAPS.

**Reply by ShaggydogUK** [reply]
My pump isn't connected to the transmiitter, CamAPS runs on the phone.  It's a closed loop system like AAPS (but different).  At present, when not playing with xDrip, I use my phone and the Dexcom receiver, which works fine.  The problem is not the number of connections, it is trying to get one phone to connect with two apps to the same transmitter, even if using both slots it will not work.  CamAPS uses Dana pumps, which do not have the capability of connecting directly to the transmitter.

**Comment by sboernemeyer** [comment]
THX Navid,
i used AAPS before :)

Mit freundlichen Grüßen

Sven Börnemeyer
Ramgestrasse 21
46145 Oberhausen
Mobil: +49 (0) 173 480 22 15
E-Mail: ***@***.***

**Comment by jamorham** [comment]
Please can any interested CamAPS users please install the Jun 2nd nightly (or any later version) and select the `Companion (testing)` data source and let me know if that works? Thanks!

**Reply by PieterDeBruijn** [reply]
Wow this is a gamechanger for CamAPS users!! Amazing, thanks!

Any chance we can also het IOB, carbs and other info?
We sometimes get gaps in the measurements.

**Reply by jrw35** [reply]
Thank you for adding this new data source for CamAPS users -- it's awesome. Setting it up was very easy, and the results are fantastic. I'm really pleased you got this approach to work.

CamAPS is a great system but I have really missed having my BG on my Garmin watch. Thank you for giving me back that feature.

**Reply by webwude** [reply]
As we do not have it yet: what is now the solution:
- accessing camapps data from the cloud or
- using both xdrip and camapps on the same phone and acessing the data directly?

**Reply by jrw35** [reply]
You need to have xDrip and CamAPS on the same phone for this to work.

As far as I know it is impossible to access the CamAPS cloud data without using Diasend.

I have xDrip connected to my Garmin watch to display the data. Or you could upload the data from xDrip to the cloud (like Nightscout) and then use that data somewhere else -- for example, on an Apple Watch that was connected to a different phone.

**Reply by Navid200** [reply]
@webwude You can enable "Companion (Testing)".  Then, xDrip will get its readings from CamAPS on the same phone.
The shortcoming of this solution is that there will be no backfill in xDrip although there may be in CamAPS.

**Reply by webwude** [reply]
Thx for the quick reply. At the moment we still have a Medtronic 770G and we upload the carelink data which is uploaded via a Medtronic apo to NightScout (with a second device). If we can stay with NightScout with using CamAPS this will be definitely an advantage.,

**Reply by funkstille** [reply]
@jamorham 
Thank you for this nice feature. Works great!

**Reply by PieterDeBruijn** [reply]
>The shortcoming of this solution is that there will be no backfill in xDrip although there may be in CamAPS.

Any news on the backfill? Is this a bug or a missing feature? I'm pretty sure the CamAPS developers are willing to work with you to create an even better experience. Let me know.

**Reply by funkstille** [reply]
As far as I understand the Companion feature in xDrip… there will be no backfill possible. Because xDrip receive the values by catching Android popup messages from CamAPS. So the source is not directly CamAPS, it’s Android. And if there was no CGM value popup because the Transmitter was out of range, there is no solution to catch the old values when the transmitter is connected again.

If the CamAPS developers really want to help they would need to implement a interface from CamAPS to xDrip. Something like it has been done within the BYODA (Dexcom app) which exposes the values locally on the Android device. I have asked for it about 6 months ago and never received an answer.

My 2 cents about CamAPS developers: There have been a couple of features that have been promised (like watch faces, follower mode, …) within support mails. None of them has been released in time / at all. Additionally there have been some unsightly bugs in latest updates. So even though the developers might be like to help, it seems like there is to much work / backlog.

**Reply by PieterDeBruijn** [reply]
CamDiab: "Unfortunately, due to regulatory reasons, we are able to work only with regulated systems."
So currently they will not actively support a connection.

**Reply by grahamjenson** [reply]
This is amazing and I love it. Hooking up CamAPS to xdrip then sending BGL readings to dexcom has let us continue to use dexcom follow to monitor and alarm on iOS!
The only issue we have run into is that occasionally readings are missing.
![image](https://user-images.githubusercontent.com/331436/180302898-ca5374cf-d8c1-44bc-b479-8f20f583d182.jpeg)

This is not huge deal but since our alarms run off Dexcom follow it means we would get an alert 5 mins early.

**Reply by Navid200** [reply]
The companion feature has no backfill capability.  Either xDrip receives the notification or not.  If it doesn't, there is no way it can get it later.

**Reply by PieterDeBruijn** [reply]
> The companion feature has no backfill capability. Either xDrip receives the notification or not. If it doesn't, there is no way it can get it later.

...without a little help from Camdiab. Everyone should try to convince them to help lend a hand. Incomplete information is worse than unofficial/unregulated/unsupported complete information. It's not that we are asking for something illegal/complicated.

I also believe there is an option to read the SMS/text messages sent by CamAPS FX and get those values & information to xDrip+ (and on to eg. Nightscout)

**Reply by GarnierM** [reply]
Bonjour,xDrip + et CamAPS cela fonctionne mais je ne sais pas trop pourquoi.Voilà ce que j'ai fait : 
- J'ai éteint mon téléphone puis je l'ai redémarré.- Je suis allé dans menu -> paramètres -> source de données matériel et là j'ai sélectionné 1 par 1 tous les choix (en pensant que "Compagnion testing était le bon choix).
- 1 -> Bluetooth Wikel
- 2 -> xBridge Wixeetc
- 3 -> ... 
A chaque fois un message d'erreur ou rien,mais à la sélection de compagnion testing (que j'avais pourtant testé avant). cela fonctionne.J'ai donc sur mon Samsung Galaxy S20 le logiciel Dexcom et xDrip+ Cela fonctionne aussi sur ma montre samsung galaxy watch (xDrip seulement)
.Je ne peux pas vous aider plus, je n'ai pas bien compris pourquoi cela a fonctionner, Bonne chance

**Reply by buxxdev** [reply]
The notifications CamAPS gives when exceeding the dexcom ranges report ≤2.2 or ≥22.3 which xdrip seems to be unable to interpret and the value is empty.  Is there anyway these values could be interpreted as the values at the top or bottom of the range?

**Reply by Navid200** [reply]
What exactly does xDrp report when that happens?

**Reply by buxxdev** [reply]
It’s just gaps in the data, no values are shown at all

**Reply by Navid200** [reply]
Thanks.  Please give us time.

**Reply by buxxdev** [reply]
Fyi here’s sample of a low…
![B4DCEFC2-0D43-41A1-A061-5D5575609280](https://user-images.githubusercontent.com/55636308/185933498-7e691665-e51d-421a-b4be-f9bbdcba2183.jpeg)

**Reply by Navid200** [reply]
Would you be able to show the xDrip logs around that time?

**Reply by buxxdev** [reply]
This was yesterday when over 22.3
![D6CCF5F2-0E05-4F01-A0F0-6906C0BC6B06](https://user-images.githubusercontent.com/55636308/186209266-7e8ad435-ec19-4cc9-9ed5-0905a7061e2a.jpeg)

**Reply by Navid200** [reply]
@majabux Would you please open an issue?
You can decide what the title should be.  I can think of:  "When CamAPS reports <2.2 or >22.3, xDrip leaves an empty gap".

I think we can fix this.  But, I don't know when, and I don't want it to be forgotten.

**Reply by buxxdev** [reply]
Yes sure no problem, thanks

**Comment by PieterDeBruijn** [comment]
The issue has been opened here:
https://github.com/NightscoutFoundation/xDrip/issues/2292

**Comment by jamorham** [comment]
Anyone interested in back-filling, please update to Oct 13th nightly and search xDrip settings for `Companion Bluetooth` and enable the feature and let me know if that makes back-filling work correctly? If there are any problems un-check that setting to revert to the standard companion collector behavior.

**Reply by buxxdev** [reply]
Happy to give it a try but could I ask how this works?  Xdrip is not connecting to the transmitter so how is Bluetooth used?
Thanks

**Reply by AliQ101** [reply]
Not sure if it should work only with CamAPS, so I tested with G6 BYODA (G6 international v.1.10.0.2. with no direct broadcast do xDrip), HTC U11 (Android 9) and there was no back-filling. For short period (2-3 readings missed) loggs showed "timestamp unrealistic" and for longer period (more than half hour) showed "did not find any matches".

![Screenshot_20221014-183855](https://user-images.githubusercontent.com/107270312/195900304-44e22148-1591-49d7-b2e1-70de0a938077.png)

**Comment by PieterDeBruijn** [comment]
We are still getting gaps, so I'm not sure if there is any difference for us:

![Screenshot 2022-10-15 12 17 25](https://user-images.githubusercontent.com/3070301/195981190-aaabbb8c-d8cf-49c7-bbac-e9fae4c15165.png)

Logs:
![WhatsApp Image 2022-10-15 at 12 20 05](https://user-images.githubusercontent.com/3070301/195981264-995f1eb6-9576-4bac-bf94-b273b52fb491.jpeg)

**Reply by jamorham** [reply]
Please could you try this again with the October 15th Nightly and let me know if that improves the situation? Thanks

**Comment by AliQ101** [comment]
With 15th Nightly it seems no progress for me, only timestamps 1970-01-01 were changes for different times also in the future. Further, companion mode seems to be less stable and stopped working when trying received data (Dex worked properly). I should switched to Dex share mode and back to start it work again. 

![Screenshot_20221015-232459 (1)](https://user-images.githubusercontent.com/107270312/196008717-896b0070-aac7-41c3-9a38-4d501780a137.png)
![Screenshot_20221016-085412](https://user-images.githubusercontent.com/107270312/196023959-e0deba65-db24-4e76-8843-0797f01a2525.png)

**Reply by PieterDeBruijn** [reply]
Same here.

**Comment by PieterDeBruijn** [comment]
@jamorham can we send logs, or test something else?

BTW Just updated to CamAPS build 155 (from 140), so that's a pretty big update. Hope there are lots of improvements.

**Reply by jamorham** [reply]
Please see the discussion here about how to submit logs: https://github.com/NightscoutFoundation/xDrip/discussions/2426 also please try the Nov 4th or later nightly to see if that helps.

**Comment by steve8x8** [comment]
Would https://bionicwookiee.com/2022/08/21/mixing-camaps-with-xdrip/ help a bit to get things sorted?
